File: hackerrank/algos/ml/predicting-house-prices/predict.py
#!/usr/bin/python

#Victor Guerra <vguerra@gmail.com>
#20141125

#https://www.hackerrank.com/challenges/predicting-house-prices

from sklearn import linear_model

(F, N) = raw_input().split()

F = int(F)
N = int(N)

features = []
prices = []
for x in range(N):
    arr = [float(a) for a in raw_input().split()]
    features.append(arr[0 : F])
    prices.append(arr[F])

#print features
clf = linear_model.LinearRegression();
clf.fit(features, prices);

T = int(raw_input())

X = [[float(a) for a in raw_input().split()] for b in range(T)]

for y in clf.predict(X):
    print y

File: hackerrank/algos/ml/predicting-house-prices/predict.cc
// Victor Guerra <vguerra@gmail.com>
// 20141119

// https://www.hackerrank.com/challenges/predicting-house-prices

#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>

#include <dlib/matrix.h>

int main() {
  unsigned F, N, T;
  std::cin >> F >> N;

  dlib::matrix<double> train_data(N, F + 1);
  dlib::matrix<double> features(N, F + 1);
  dlib::matrix<double> featuresT(F + 1, N);
  dlib::matrix<double, 0, 1> prices(N);
  dlib::matrix<double, 0, 1> parameters(F + 1);
  
  std::cin >> train_data;

  prices = dlib::colm(train_data, F);

  dlib::set_colm(features, 0) = 1.0;

  dlib::set_subm(features, dlib::range(0, N - 1), dlib::range(1, F - 1)) =
      dlib::subm(train_data, dlib::rectangle(0, 0, F - 1, N - 1));

  featuresT = dlib::trans(features);

  std::cin >> T;
  dlib::matrix<double> features_test(T, F + 1);
  dlib::matrix<double> test_data(T, F);
  std::cin >> test_data;

  dlib::set_colm(features_test, 0) = 1.0;
  dlib::set_subm(features_test, dlib::range(0, T - 1), dlib::range(1, F - 1)) =
      test_data;
  //std::cout << features_test;
  parameters = dlib::inv(featuresT * features) * featuresT * prices;

  //std::cout << parameters;
  std::cout << features_test*parameters;

  return 0;
}
